core/tee/tee_rpmb_fs.c

```diff
@@ -1472,54 +1503,325 @@ static TEE_Result tee_rpmb_get_max_block(uint16_t dev_id, uint32_t *max_block)
  */
 
 static TEE_Result get_fat_start_address(uint32_t *addr);
+static TEE_Result rpmb_fs_setup(void);
 
-#if (TRACE_LEVEL >= TRACE_FLOW)
-static void dump_fat(void)
+/**
+ * fat_entry_dir_free: Free the FAT entry dir.
+ */
+static void fat_entry_dir_free(void)
+{
+	if (fat_entry_dir) {
+		free(fat_entry_dir->rpmb_fat_entry_buf);
+		free(fat_entry_dir);
+		fat_entry_dir = NULL;
+	}
+}
+
+/**
+ * fat_entry_dir_init: Initialize the FAT FS entry buffer/cache
+ * This function must be called before reading FAT FS entries using the
+ * function fat_entry_dir_get_next. This initializes the buffer/cache with the
+ * first FAT FS entries, i.e., 'CFG_RPMB_FS_RD_ENTRIES' when caching is
+ * disabled and MIN(CFG_RPMB_FS_CACHE_ENTRIES, CFG_RPMB_FS_RD_ENTRIES) when
+ * caching is enabled.
+ */
+static TEE_Result fat_entry_dir_init(void)
 {
 	TEE_Result res = TEE_ERROR_GENERIC;
-	struct rpmb_fat_entry *fat_entries = NULL;
-	uint32_t fat_address;
-	size_t size;
-	int i;
-	bool last_entry_found = false;
+	struct rpmb_fat_entry *fe = NULL;
+	uint32_t fat_address = 0;
+	uint32_t num_elems_read = 0;
+
+	if (fat_entry_dir)
+		return TEE_SUCCESS;
+
+	res = rpmb_fs_setup();
+	if (res)
+		return res;
 
 	res = get_fat_start_address(&fat_address);
-	if (res != TEE_SUCCESS)
-		goto out;
+	if (res)
+		return res;
+
+	fat_entry_dir = calloc(1, sizeof(struct rpmb_fat_entry_dir));
+	if (!fat_entry_dir)
+		return TEE_ERROR_OUT_OF_MEMORY;
 
-	size = CFG_RPMB_FS_RD_ENTRIES * sizeof(struct rpmb_fat_entry);
-	fat_entries = malloc(size);
-	if (!fat_entries) {
+	/*
+	 * If caching is enabled, read in up to the maximum cache size, but
+	 * never more than the single read in size. Otherwise, read in as many
+	 * entries fit into the temporary buffer.
+	 */
+	if (CFG_RPMB_FS_CACHE_ENTRIES)
+		num_elems_read = MIN(CFG_RPMB_FS_CACHE_ENTRIES,
+				     CFG_RPMB_FS_RD_ENTRIES);
+	else
+		num_elems_read = CFG_RPMB_FS_RD_ENTRIES;
+
+	/*
+	 * Allocate memory for the FAT FS entries to read in.
+	 */
+	fe = calloc(num_elems_read, sizeof(struct rpmb_fat_entry));
+	if (!fe) {
 		res = TEE_ERROR_OUT_OF_MEMORY;
 		goto out;
 	}
 
-	while (!last_entry_found) {
-		res = tee_rpmb_read(CFG_RPMB_FS_DEV_ID, fat_address,
-				    (uint8_t *)fat_entries, size, NULL, NULL);
-		if (res != TEE_SUCCESS)
-			goto out;
+	res = tee_rpmb_read(CFG_RPMB_FS_DEV_ID, fat_address, (uint8_t *)fe,
+			    num_elems_read * sizeof(*fe), NULL, NULL);
+	if (res)
+		goto out;
```

Comment:
A RPMB enrty is 512 byte which is not negligible.
If `tee_rpmb_read()` fails because of heap issues, maybe it should try again without filling the cache, just for the initial request entry/entries.

Reply:
At this point, we do not necessarily fill the cache, but may just initialize the temp buffer if caching is disabled.
If caching is disabled, we try to read in CFG_RPMB_FS_RD_ENTRIES entries.
If caching is enabled, we try to read in CFG_RPMB_FS_RD_ENTRIES or CFG_RPMB_FS_CACHE_ENTRIES entries whichever is smaller.
Meaning, caching can even reduce the cost of memory at this point of initialization, compared to the previous implementation.

Further, caching is an optional feature and the previously introduced variable 'CFG_RPMB_FS_RD_ENTRIES' to adjust the read in size is an optional adjustment as well.
From that perspective, 'nothing has changed' compared to prior implementations.
However, people can adjust the settings/behavior to improve RPMB I/O while being aware of the available memory on their platform and of the consequences of modifying these parameters.
I think, that is a pretty reasonable assumption.

In general, I'm not saying we don't want a retry here. I'd say if we want a retry functionality here, I would prefer to address this in a separate patch and to conclude this one.
Also, there are more memory allocations when using the cache (in fat_entry_dir_get_next), so there might be more places in the code where we'd place a retry.